File: src/decoder.ts
export type OnData = (payloadFragment: string) => void;

/**
 * Streaming chunked decoder for the simplified “problem set” format:
 *
 *   <hex-size>\r\n<payload>\r\n ... 0\r\n\r\n
 *
 * Assumptions (matches the prompt examples):
 * - Input arrives as JS strings.
 * - “Size” counts JS characters (ASCII payload). This is NOT byte-accurate for UTF-8/binary.
 * - No chunk extensions (e.g. ";ext=...") and no trailers.
 */
export class ChunkedDecoder {
  private state: "SIZE" | "PAYLOAD" | "EXPECT_CRLF" | "DONE" = "SIZE";

  // SIZE parsing (tiny state)
  private sizeHex = "";
  private sawCR = false;

  // PAYLOAD parsing
  private remaining = 0;

  // CRLF expectation parsing
  private expectIndex = 0; // 0 => expect '\r', 1 => expect '\n'
  private afterExpect: "SIZE" | "DONE" = "SIZE";

  constructor(private readonly onData: OnData) {}

  decodeChunk(chunk: string): void {
    if (this.state === "DONE") return;

    // Cursor into the current chunk
    let i = 0;
    while (i < chunk.length) {

      // STATE: gathering the hex size line
      if (this.state === "SIZE") {
        const c = chunk[i++];

        // If we previously consumed a '\r' for the size line, the next char MUST be '\n'.
        // 
        if (this.sawCR) {
          if (c !== "\n") throw new Error("Invalid chunked encoding: expected LF after CR in size line.");
          this.sawCR = false;

          const n = Number.parseInt(this.sizeHex.trim() || "0", 16);
          if (!Number.isFinite(n) || n < 0) throw new Error(`Invalid chunk size: "${this.sizeHex}"`);

          this.sizeHex = "";
          this.remaining = n;

          if (n === 0) {
            // Simplified termination: after "0\r\n" we expect the final "\r\n".
            this.startExpectCRLF("DONE");
          } else {
            this.state = "PAYLOAD";
          }
          continue;
        }

        // Found the end of the size line!
        if (c === "\r") {
          this.sawCR = true;
          continue;
        }

        // Assumption: valid stream, no extensions. Collect the size line verbatim until CR.
        this.sizeHex += c;
        continue;
      }

      // STATE: consuming payload by known size
      if (this.state === "PAYLOAD") {
        // Greedily consume payload from the current fragment without buffering.
        const available = chunk.length - i;
        const take = Math.min(this.remaining, available);

        if (take > 0) {
          this.onData(chunk.slice(i, i + take));
          i += take;
          this.remaining -= take;
        }

        if (this.remaining === 0) {
          // After payload there must be a CRLF terminator.
          this.startExpectCRLF("SIZE");
        }
        continue;
      }

      // STATE: expecting either payload CRLF (1) or final CRLF (2).
      // Moves either to SIZE (1) or DONE (2) state.
      if (this.state === "EXPECT_CRLF") {
        const expected = this.expectIndex === 0 ? "\r" : "\n";
        const c = chunk[i++];

        if (c !== expected) throw new Error("Invalid chunked encoding: expected CRLF.");

        this.expectIndex++;
        if (this.expectIndex === 2) {
          this.expectIndex = 0;
          this.state = this.afterExpect === "DONE" ? "DONE" : "SIZE";
          if (this.state === "DONE") return;
        }
        continue;
      }

      // DONE
      return;
    }
  }

  isDone(): boolean {
    return this.state === "DONE";
  }

  /** Throws unless we have consumed a full terminal 0-sized chunk (0\r\n\r\n). */
  finalize(): void {
    if (!this.isDone()) throw new Error("Chunked stream not finished.");
  }

  private startExpectCRLF(next: "SIZE" | "DONE"): void {
    this.state = "EXPECT_CRLF";
    this.expectIndex = 0;
    this.afterExpect = next;
  }
}

/**
 * A string accumulator that avoids pathological memory churn when the stream is split into
 * extremely tiny fragments. It groups many small fragments into medium-sized blocks.
 *
 * - Each fragment is stored once in `pending`.
 * - Occasionally, `pending` is joined into a single block string.
 * - At the very end, all blocks are joined to produce one final string.
 *
 * This ensures any character is copied O(1) times: once into a block and once
 * into the final result.
 */
class StringAccumulator {
  private blocks: string[] = [];
  private pending: string[] = [];
  private pendingChars = 0;

  constructor(
    private readonly maxPendingChars = 64 * 1024,
    private readonly maxPendingParts = 2048
  ) {}

  push(fragment: string): void {
    if (fragment.length === 0) return;

    this.pending.push(fragment);
    this.pendingChars += fragment.length;

    if (this.pendingChars >= this.maxPendingChars || this.pending.length >= this.maxPendingParts) {
      this.flush();
    }
  }

  flush(): void {
    if (this.pending.length === 0) return;
    this.blocks.push(this.pending.join(""));
    this.pending = [];
    this.pendingChars = 0;
  }

  toString(): string {
    if (this.blocks.length === 0) {
      // Avoid one extra join in the common case.
      if (this.pending.length === 0) return "";
      if (this.pending.length === 1) return this.pending[0];
      return this.pending.join("");
    }

    // Flush pending into blocks once, then do a single final join.
    this.flush();

    if (this.blocks.length === 1) return this.blocks[0];
    return this.blocks.join("");
  }
}

/**
 * Convenience wrapper matching the “decoder with a result” style.
 *
 * Use ChunkedDecoder directly if you want true streaming output.
 */
export class ChunkedCollectingDecoder {
  private readonly accumulator = new StringAccumulator();
  private readonly decoder = new ChunkedDecoder((s) => this.accumulator.push(s));

  decodeChunk(chunk: string): void {
    this.decoder.decodeChunk(chunk);
  }

  finalize(): void {
    this.decoder.finalize();
  }

  get result(): string {
    // ! Allow partial results
    // if (!this.decoder.isDone()) {
    //   throw new Error("Not finished yet (call finalize() after feeding all chunks).");
    // }
    return this.accumulator.toString();
  }
}

// Backwards-compatible alias used by tests/scripts/docs.
export { ChunkedCollectingDecoder as CollectingDecoder };

File: rust/decoder.rs
//! Rust implementation of the simplified streaming chunked decoder used in this repo.
//!
//! Framing:
//!   `<hex-size>\r\n<payload>\r\n ... 0\r\n\r\n`
//!
//! Matches the TypeScript version (`src/decoder.ts`):
//! - Input is provided as text fragments (`&str`), assumed to be ASCII.
//! - Chunk size counts characters (ASCII => 1 char == 1 byte).
//! - No chunk extensions and no trailers.

use std::error::Error;
use std::fmt;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum DecodeError {
    NonAsciiInput,
    ExpectedLfAfterCrInSizeLine,
    EmptyChunkSize,
    InvalidChunkSize,
    ChunkSizeOverflow,
    ExpectedCrlf,
    NotFinished,
}

impl fmt::Display for DecodeError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            DecodeError::NonAsciiInput => write!(
                f,
                "Non-ASCII input is not supported (this repo assumes ASCII payload)."
            ),
            DecodeError::ExpectedLfAfterCrInSizeLine => write!(
                f,
                "Invalid chunked encoding: expected LF after CR in size line."
            ),
            DecodeError::EmptyChunkSize => write!(f, "Invalid chunk size: \"\""),
            DecodeError::InvalidChunkSize => write!(f, "Invalid chunk size."),
            DecodeError::ChunkSizeOverflow => write!(f, "Invalid chunk size (overflow)."),
            DecodeError::ExpectedCrlf => write!(f, "Invalid chunked encoding: expected CRLF."),
            DecodeError::NotFinished => write!(f, "Chunked stream not finished."),
        }
    }
}

impl Error for DecodeError {}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum State {
    Size,
    Payload,
    ExpectCrlf,
    Done,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum AfterExpect {
    Size,
    Done,
}

fn hex_value(b: u8) -> Option<u8> {
    match b {
        b'0'..=b'9' => Some(b - b'0'),
        b'a'..=b'f' => Some(b - b'a' + 10),
        b'A'..=b'F' => Some(b - b'A' + 10),
        _ => None,
    }
}

/// Streaming decoder state machine (greedy, minimal buffering).
#[derive(Debug)]
pub struct ChunkedDecoder {
    state: State,

    // SIZE parsing
    size_acc: usize,
    size_digits: usize,
    saw_cr: bool,

    // PAYLOAD parsing
    remaining: usize,

    // CRLF expectation parsing
    expect_index: u8, // 0 => expect '\r', 1 => expect '\n'
    after_expect: AfterExpect,
}

impl ChunkedDecoder {
    pub fn new() -> Self {
        Self {
            state: State::Size,
            size_acc: 0,
            size_digits: 0,
            saw_cr: false,
            remaining: 0,
            expect_index: 0,
            after_expect: AfterExpect::Size,
        }
    }

    pub fn decode_chunk<F>(&mut self, chunk: &str, mut on_data: F) -> Result<(), DecodeError>
    where
        F: FnMut(&str),
    {
        if self.state == State::Done {
            return Ok(());
        }

        if !chunk.is_ascii() {
            return Err(DecodeError::NonAsciiInput);
        }

        let bytes = chunk.as_bytes();
        let mut i = 0usize;
        while i < bytes.len() {
            match self.state {
                State::Size => {
                    let b = bytes[i];
                    i += 1;

                    if self.saw_cr {
                        if b != b'\n' {
                            return Err(DecodeError::ExpectedLfAfterCrInSizeLine);
                        }
                        self.saw_cr = false;

                        if self.size_digits == 0 {
                            return Err(DecodeError::EmptyChunkSize);
                        }

                        let n = self.size_acc;
                        self.size_acc = 0;
                        self.size_digits = 0;
                        self.remaining = n;

                        if n == 0 {
                            self.start_expect_crlf(AfterExpect::Done);
                        } else {
                            self.state = State::Payload;
                        }
                        continue;
                    }

                    if b == b'\r' {
                        self.saw_cr = true;
                        continue;
                    }

                    let Some(d) = hex_value(b) else {
                        return Err(DecodeError::InvalidChunkSize);
                    };

                    self.size_acc = self
                        .size_acc
                        .checked_mul(16)
                        .and_then(|v| v.checked_add(d as usize))
                        .ok_or(DecodeError::ChunkSizeOverflow)?;
                    self.size_digits += 1;
                }
                State::Payload => {
                    let available = bytes.len() - i;
                    let take = self.remaining.min(available);

                    if take > 0 {
                        // ASCII => byte indices are valid UTF-8 boundaries.
                        on_data(&chunk[i..i + take]);
                        i += take;
                        self.remaining -= take;
                    }

                    if self.remaining == 0 {
                        self.start_expect_crlf(AfterExpect::Size);
                    }
                }
                State::ExpectCrlf => {
                    let expected = if self.expect_index == 0 { b'\r' } else { b'\n' };
                    let b = bytes[i];
                    i += 1;

                    if b != expected {
                        return Err(DecodeError::ExpectedCrlf);
                    }

                    self.expect_index += 1;
                    if self.expect_index == 2 {
                        self.expect_index = 0;
                        self.state = match self.after_expect {
                            AfterExpect::Done => State::Done,
                            AfterExpect::Size => State::Size,
                        };
                        if self.state == State::Done {
                            return Ok(());
                        }
                    }
                }
                State::Done => return Ok(()),
            }
        }

        Ok(())
    }

    pub fn is_done(&self) -> bool {
        self.state == State::Done
    }

    pub fn finalize(&self) -> Result<(), DecodeError> {
        if self.is_done() {
            Ok(())
        } else {
            Err(DecodeError::NotFinished)
        }
    }

    fn start_expect_crlf(&mut self, next: AfterExpect) {
        self.state = State::ExpectCrlf;
        self.expect_index = 0;
        self.after_expect = next;
    }
}

impl Default for ChunkedDecoder {
    fn default() -> Self {
        Self::new()
    }
}

/// Convenience wrapper matching the prompt-style `Decoder { decodeChunk, result }`.
#[derive(Debug, Default)]
pub struct ChunkedCollectingDecoder {
    decoder: ChunkedDecoder,
    result: String,
}

impl ChunkedCollectingDecoder {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn decode_chunk(&mut self, chunk: &str) -> Result<(), DecodeError> {
        let out = &mut self.result;
        self.decoder.decode_chunk(chunk, |frag| out.push_str(frag))
    }

    pub fn is_done(&self) -> bool {
        self.decoder.is_done()
    }

    pub fn finalize(&self) -> Result<(), DecodeError> {
        self.decoder.finalize()
    }

    pub fn result(&self) -> &str {
        &self.result
    }
}

/// Backwards-compatible alias.
pub type CollectingDecoder = ChunkedCollectingDecoder;

/// Alias for the prompt's class name.
pub type Decoder = ChunkedCollectingDecoder;

#[cfg(test)]
mod tests {
    use super::*;

    fn encode_chunked(payload: &str, chunk_sizes: &[usize]) -> String {
        assert!(payload.is_ascii());

        let bytes = payload.as_bytes();
        let mut i = 0usize;
        let mut out = String::new();

        for &size in chunk_sizes {
            if i >= bytes.len() {
                break;
            }
            let take = size.min(bytes.len() - i);
            out.push_str(&format!("{:x}\r\n", take));
            out.push_str(std::str::from_utf8(&bytes[i..i + take]).unwrap());
            out.push_str("\r\n");
            i += take;
        }

        if i < bytes.len() {
            let take = bytes.len() - i;
            out.push_str(&format!("{:x}\r\n", take));
            out.push_str(std::str::from_utf8(&bytes[i..]).unwrap());
            out.push_str("\r\n");
        }

        out.push_str("0\r\n\r\n");
        out
    }

    #[test]
    fn decodes_across_arbitrary_fragmentation() {
        let payload = "Hello, world! This is chunked.";
        let encoded = encode_chunked(payload, &[1, 2, 3, 4, 5]);

        let mut out = String::new();
        let mut decoder = ChunkedDecoder::new();

        for frag in encoded.as_bytes().chunks(3) {
            let frag = std::str::from_utf8(frag).unwrap();
            decoder.decode_chunk(frag, |d| out.push_str(d)).unwrap();
        }
        decoder.finalize().unwrap();

        assert_eq!(out, payload);
    }

    #[test]
    fn collecting_decoder_matches_streaming_output() {
        let payload = "abcdefg";
        let encoded = encode_chunked(payload, &[2, 2, 3]);

        let mut dec = ChunkedCollectingDecoder::new();
        for frag in encoded.as_bytes().chunks(2) {
            dec.decode_chunk(std::str::from_utf8(frag).unwrap()).unwrap();
        }
        dec.finalize().unwrap();
        assert_eq!(dec.result(), payload);
    }

    #[test]
    fn payload_can_contain_crlf() {
        let payload = "a\r\nb\r\nc";
        let encoded = encode_chunked(payload, &[2, 2, 2]);

        let mut dec = ChunkedCollectingDecoder::new();
        for frag in encoded.as_bytes().chunks(1) {
            dec.decode_chunk(std::str::from_utf8(frag).unwrap()).unwrap();
        }
        dec.finalize().unwrap();
        assert_eq!(dec.result(), payload);
    }
}
